### PR TITLE
feat: metadata endpoint

### PR DESCRIPTION
**Pull request by preschian**
Fix: #155
Fix: #95

Related changes on our UI: https://github.com/kodadot/nft-gallery/pull/8171

<details><summary>test case:</summary>
<p>

return json
- http://127.0.0.1:8787/metadata?url=https://image-beta.w.kodadot.xyz/ipfs/bafkreia3j75r474kgxxmptwh5n43j5nrvn3du5l7dcfq2twh73wmagqs6m
- http://127.0.0.1:8787/metadata?url=https://cloudflare-ipfs.com/ipfs/bafkreia3j75r474kgxxmptwh5n43j5nrvn3du5l7dcfq2twh73wmagqs6m

return json from non ipfs
- http://127.0.0.1:8787/metadata?url=https://polkadot-data.s3.us-east-2.amazonaws.com/metadata/nfts-88/nfts-88_collection-meta.json
- http://127.0.0.1:8787/metadata?url=https://image.w.kodadot.xyz/type/url?endpoint=https://polkadot-data.s3.us-east-2.amazonaws.com/metadata/nfts-88/nfts-88_collection-meta.json

redirect
- http://127.0.0.1:8787/metadata?url=https://image-beta.w.kodadot.xyz/ipfs/bafybeihpbyskcjjhb5tnrjbsnsvzcfm2npjec3uf65n3o7stucta2rlos4
- http://127.0.0.1:8787/metadata?url=https://cloudflare-ipfs.com/ipfs/bafybeihpbyskcjjhb5tnrjbsnsvzcfm2npjec3uf65n3o7stucta2rlos4

redirect from non ipfs
- http://127.0.0.1:8787/metadata?url=https://polkadot-data.s3.us-east-2.amazonaws.com/metadata/nfts-88/nfts-88_collection-img.png
- http://127.0.0.1:8787/metadata?url=https://image.w.kodadot.xyz/type/url?endpoint=https://polkadot-data.s3.us-east-2.amazonaws.com/metadata/nfts-88/nfts-88_collection-img.png

</p>
</details>

### CODE REVIEW

File: image/src/routes/metadata.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+type HonoInterface = Context<
+  {
+    Bindings: Env
+  },
+  '/metadata',
+  {}
+>
+
+const getMimeType = async (url: string) => {
```
Comment (vikiival) [inline image/src/routes/metadata.ts:16]: This is already a part of minipfs. 🤗

File: image/src/routes/metadata.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    const metadataKV = await c.env.METADATA.get(key)
+    if (metadataKV) {
+      return c.json(JSON.parse(metadataKV))
+    }
+
+    // 1. put to KV
+    // ----------------------------------------
+    const externalUrl = url.includes('kodadot.xyz') ? toNftstorage(url) : url
+    const data = await fetch(externalUrl)
+    const json = await data.json()
```
Comment (vikiival) [inline image/src/routes/metadata.ts:56]: Json takes generic as json<>

File: image/src/routes/metadata.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    const animationUrl = normalized.animationUrl
+    let animationUrlMetadata
+    if (animationUrl) {
+      animationUrlMetadata = await getMimeType(animationUrl)
+    }
+
+    const predefinedAttributes = {
+      image,
+      imageMetadata,
```
Comment (vikiival) [inline image/src/routes/metadata.ts:72]: imageMimeType

File: image/src/routes/metadata.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    const json = await data.json()
+    const content = contentFrom(json as BaseMetadata)
+    // @ts-ignore
+    const normalized = normalize(content, ipfsUrl)
+
+    const image = normalized.image
+    const imageMetadata = await getMimeType(image)
+
+    const animationUrl = normalized.animationUrl
+    let animationUrlMetadata
```
Comment (vikiival) [inline image/src/routes/metadata.ts:65]: mimeType

File: image/src/routes/metadata.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    if (animationUrl) {
+      animationUrlMetadata = await getMimeType(animationUrl)
```
Comment (vikiival) [inline image/src/routes/metadata.ts:67]: It can be present in metadata under `type`

**Review by vikiival — COMMENTED** [review]
Like good solution but code is very hard to read. 

Is is possible to put code repetitions in the packages?

**Comment by preschian** [comment]
> Is is possible to put code repetitions in the packages?

actually still not ready to review, which section is repetitions?

**Comment by preschian** [comment]
> Like good solution but code is very hard to read.

which section is hard to read? I can improve. the code should be straightforward
```
if kv -> return from kv
else construct metadata
```

**Comment by preschian** [comment]
please ignore `externalUrl` section, I'm currently still not tested this one https://community.cloudflare.com/t/get-error-code-1042-when-fetching-within-worker/288031

File: image/src/routes/metadata.ts [hunk 6]
```diff
@@ -59,19 +59,19 @@ export const getMetadata = async (c: HonoInterface) => {
...
+      animationUrlMimeType = await getMimeType(animationUrl)
     }
 
     const predefinedAttributes = {
       image,
-      imageMetadata,
+      imageMimeType,
       animationUrl,
-      animationUrlMetadata,
+      animationUrlMimeType,
```
Comment (vikiival) [inline image/src/routes/metadata.ts:74]:
```suggestion
      mimeType: animationUrlMimeType,
```

File: image/src/routes/metadata.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    if (metadataKV) {
+      return c.json(JSON.parse(metadataKV))
+    }
+
+    // 1. put to KV
+    // ----------------------------------------
+    const externalUrl = url.includes('kodadot.xyz') ? toNftstorage(url) : url // TODO: probably not needed
+    const data = await fetch(externalUrl)
+    const json = await data.json()
+    const content = contentFrom(json as BaseMetadata)
```
Comment (vikiival) [inline image/src/routes/metadata.ts:57]:
```suggestion
    const json = await data.json<BaseMetadata>()
    const content = contentFrom(json)
```

File: image/src/routes/metadata.ts (hunk 2 above)
Comment (vikiival) [reply image/src/routes/metadata.ts:56]: https://github.com/kodadot/workers/pull/206/files#r1399160973

File: image/src/routes/metadata.ts (hunk 6 above)
Comment (preschian) [reply image/src/routes/metadata.ts:74]: we need both mimeType actually, for image and animationUrl

File: image/src/utils/ipfs.ts [hunk 8]
```diff
@@ -1,9 +1,28 @@
...
+  return url.toString()
+}
+
+export function ipfsUrl(ipfs?: string) {
+  if (!ipfs) {
+    return ''
+  }
+
+  // TODO: 'kodadot_beta' for beta, 'kodadot' for prod
+  return $purify(ipfs, ['kodadot_beta'])[0]
```
Comment (vikiival) [inline image/src/utils/ipfs.ts:23]:
```suggestion
  return purifySingle(ipfs, 'kodadot_beta')
```

Comment (vikiival) [reply image/src/utils/ipfs.ts:23]: Needs to be implemented on the side of `minipfs`

File: image/src/routes/metadata.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    if (!image.includes('w.kodadot.xyz')) {
+      predefinedAttributes.image = toCustomKodaURL(image)
+    }
+
+    if (!animationUrl.includes('w.kodadot.xyz')) {
+      predefinedAttributes.animationUrl = toCustomKodaURL(animationUrl)
+    }
+
+    const attributes = { ...predefinedAttributes, _raw: normalized }
+
```
Comment (vikiival) [inline image/src/routes/metadata.ts:89]: What does this app do?

**Comment by vikiival** [comment]
> which section is hard to read? I can improve

https://github.com/kodadot/workers/pull/206/files#r1399166160

File: image/src/routes/metadata.ts (hunk 9 above)
Comment (preschian) [reply image/src/routes/metadata.ts:89]: Ignore most parts of this section. most of the attributes should be from `hyperdata`. extended attributes from this worker should be `mimeType` and `thumbnail (once we already have cf-streams)`

**Comment by preschian** [comment]
@vikiival After some testing, I will keep this part https://github.com/kodadot/workers/blob/5aa42c7d4f87eb17a2e2ae10051e3325ce804c01/image/src/routes/metadata.ts#L21-L35

1. throw error if using `obtainMimeType` -> `invalid metadata json TypeError: Cannot destructure property 'type' of '(intermediate value)' as it is undefined.`
2. if you check `nft.type` in this issue is invalid: https://github.com/kodadot/workers/issues/95. so, check proper mime type is required
3. I just realized that sometimes gateway not return proper mimetype, so I need to recheck again with `if (data.status !== 200)` in that part

File: image/src/utils/ipfs.ts [hunk 10]
```diff
@@ -1,9 +1,28 @@
...
   gateway1: string
   gateway2: string
 }
 
+export function toNftstorage(path: string) {
+  const infura = new URL(getProviderList(['nftstorage'])[0])
+  const url = new URL(path)
+  url.hostname = infura.hostname
+
+  return url.toString()
```
[the remark this answers is not in the document]
Comment (preschian) [reply image/src/utils/ipfs.ts]: not needed for now, because still achievable with native URL API

File: image/src/utils/ipfs.ts (hunk 8 above)
Comment (preschian) [reply image/src/utils/ipfs.ts:23]: for `minipfs` I created this issue for followup PRs https://github.com/kodadot/packages/issues/182

**Comment by preschian** [comment]
@vikiival ok, now ready to review, lmk if you find still hard to read

**Comment by vikiival** [comment]
> if you check nft.type in this issue is invalid: https://github.com/kodadot/workers/issues/95. so, check proper mime type is required

It is tested in the hyperdata

https://github.com/kodadot/packages/blob/4774917503b45bd71a831b78ebba12a167adf69f/hyperdata/src/normalize.ts#L63

**Comment by preschian** [comment]
> > if you check nft.type in this issue is invalid: #95. so, check proper mime type is required
> 
> It is tested in the hyperdata
> 
> [kodadot/packages@`4774917`/hyperdata/src/normalize.ts#L63](https://github.com/kodadot/packages/blob/4774917503b45bd71a831b78ebba12a167adf69f/hyperdata/src/normalize.ts#L63)

Yes, Then hyperdata will return an empty string. Meanwhile, it would be good if we returned this to our UI. End goals: it would be even better if we could put that in our indexer

for comparison, I expect something like `primary_mime_type` [from this API](https://instasearch.rmrk.link/collections/nfts/documents/search?q=*&filter_by=is_burned:+false+%26%26+is_hidden:+!%3D+true+%26%26+(primary_media_type:+%3D+[image,gif,video,model,audio,pdf,unknown]+%7C%7C+primary_asset_primary_media_type:+%3D+[image,gif,video,model,audio,pdf,unknown])+%26%26+is_verified:+true+%26%26+(primary_media_type:+%3D+[video]+%7C%7C+primary_asset_primary_media_type:+%3D+[video])&query_by=name,+primary_asset_name,+collection_name&per_page=36&page=1&sort_by=minted_at:desc&group_by=collection_id&group_limit=1&x-typesense-api-key=vW9DI3k6ZBM7eH3TgCyyInpG2YtcaLAB)

**Comment by preschian** [comment]
> What is proper mime type?

oh, I mean check the mime type again in this worker

**Pull request merged**